File: app/js/repomap.js
(function () {
    'use strict';

    const Store = window.NymbotStore;
    const GitApi = window.NymbotGitApi;

    const CACHE_KEY = 'map_repos';
    const FRESH_MS = 2 * 3600 * 1000;
    const RETRY_MS = 5 * 60 * 1000;
    const KEEP_MAPS = 12;
    const FILES_PER_REPO = 500;
    const DIR_NAMES_MAX = 24;
    const BLOCK_MAX = 12000;
    const READY_MS = 2500;

    const SKIP_DIRS = new Set([
        'node_modules', '.git', 'dist', 'build', 'out', 'target', '.next',
        '.nuxt', '.svelte-kit', '.cache', '.parcel-cache', 'coverage',
        '__pycache__', '.venv', 'venv', 'Pods', '.dart_tool', '.gradle',
        '.idea', '.vscode', 'bower_components', '.terraform', 'DerivedData',
        '.pub-cache', '.mypy_cache', '.pytest_cache', '.tox', '.expo', '.angular'
    ]);

    const SKIP_FILES = new Set([
        'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml', 'npm-shrinkwrap.json',
        'pubspec.lock', 'Cargo.lock', 'composer.lock', 'Gemfile.lock',
        'poetry.lock', 'go.sum', 'Podfile.lock', 'mix.lock', 'flake.lock',
        '.DS_Store'
    ]);

    const STOP_TERMS = new Set(('a an and are as at be but by can could did do does '
        + 'file files for from had has have how i if in is it its me my not of on or '
        + 'our so than that the their them then there these they this to was we were '
        + 'what when where which who why will with would you your please').split(' '));

    function keyFor(repo) {
        return [
            repo.provider || 'github',
            repo.host || '',
            repo.repo || '',
            repo.branch || '',
            repo.paths || ''
        ].join('|');
    }

    let memo = null;

    function held() {
        if (memo) return memo;
        const all = Store.read(CACHE_KEY, {});
        memo = all && typeof all === 'object' ? all : {};
        return memo;
    }

    function store(all) {
        memo = all;
        Store.write(CACHE_KEY, all);
    }

    function keep(all) {
        const keys = Object.keys(all)
            .sort((a, b) => (all[b].usedAt || all[b].at || 0) - (all[a].usedAt || all[a].at || 0))
            .slice(0, KEEP_MAPS);
        const out = {};
        for (const k of keys) out[k] = all[k];
        store(out);
        return out;
    }

    function scopes(repo) {
        return String(repo.paths || '').split(/[,\s]+/).filter(Boolean)
            .map(p => p.replace(/^\/+/, ''));
    }

    function inScope(path, only) {
        if (!only.length) return true;
        for (const p of only) {
            if (path === p) return true;
            if (path.indexOf(p.endsWith('/') ? p : p + '/') === 0) return true;
        }
        return false;
    }

    function ignorable(path) {
        const parts = path.split('/');
        if (SKIP_FILES.has(parts[parts.length - 1])) return true;
        for (let i = 0; i < parts.length - 1; i++) {
            if (SKIP_DIRS.has(parts[i])) return true;
        }
        return false;
    }

    function depthOf(path) {
        let n = 0;
        for (let i = 0; i < path.length; i++) if (path[i] === '/') n++;
        return n;
    }

    function usable(paths, repo) {
        const only = scopes(repo || {});
        const rows = [];
        for (const raw of paths || []) {
            const path = typeof raw === 'string' ? raw : '';
            if (!path || path.endsWith('/')) continue;
            if (ignorable(path)) continue;
            if (!inScope(path, only)) continue;
            rows.push(path);
        }
        rows.sort((a, b) => depthOf(a) - depthOf(b) || a.localeCompare(b));
        const kept = rows.slice(0, FILES_PER_REPO);
        kept.sort((a, b) => a.localeCompare(b));
        return { paths: kept, dropped: rows.length - kept.length, total: rows.length };
    }

    function termsOf(query) {
        return new Set(String(query || '').toLowerCase().split(/[^a-z0-9]+/)
            .filter(w => w.length > 2 && !STOP_TERMS.has(w)));
    }

    function wanted(name, terms) {
        if (!terms.size) return false;
        const plain = name.toLowerCase();
        for (const term of terms) {
            if (plain.indexOf(term) !== -1) return true;
        }
        return false;
    }

    function group(paths) {
        const dirs = new Map();
        for (const path of paths) {
            const at = path.lastIndexOf('/');
            const dir = at === -1 ? '' : path.slice(0, at + 1);
            const name = at === -1 ? path : path.slice(at + 1);
            if (!dirs.has(dir)) dirs.set(dir, []);
            dirs.get(dir).push(name);
        }
        return dirs;
    }

    function heading(repo, entry) {
        const branch = entry.branch ? '@' + entry.branch : '';
        const scope = repo.paths ? ', paths ' + repo.paths : '';
        const plural = entry.total === 1 ? ' file' : ' files';
        return `--- ${repo.repo}${branch} (${entry.total}${plural}${scope}) ---`;
    }

    function render(repo, entry, budget, terms) {
        const lines = [heading(repo, entry)];
        let spent = lines[0].length;
        const dirs = group(entry.paths || []);
        const names = Array.from(dirs.keys()).sort();
        let skippedFiles = 0;

        for (const dir of names) {
            const all = dirs.get(dir).sort();
            if (spent >= budget) {
                skippedFiles += all.length;
                continue;
            }
            const first = all.filter(n => wanted(n, terms));
            const rest = all.filter(n => !wanted(n, terms));
            const ordered = first.concat(rest);
            const shown = ordered.slice(0, DIR_NAMES_MAX);
            const over = ordered.length - shown.length;
            const line = (dir || '(root)') + ': ' + shown.join(', ')
                + (over > 0 ? `, and ${over} more here` : '');
            spent += line.length + 1;
            lines.push(line);
        }

        const unlisted = (entry.dropped || 0) + skippedFiles;
        if (unlisted) {
            lines.push(`${unlisted} further files are not named above — list a directory to see them.`);
        }
        if (entry.partial) {
            lines.push('The forge cut this listing short, so it is not every file.');
        }
        return lines.join('\n');
    }

    const pending = new Map();

    let lastBlock = { sig: '', text: '' };

    const RepoMap = {
        keyFor,

        entry(repo) {
            const all = held();
            const hit = all[keyFor(repo)];
            return hit && Array.isArray(hit.paths) ? hit : null;
        },

        stale(repo) {
            const hit = this.entry(repo);
            if (!hit) return true;
            const age = Date.now() - (hit.at || 0);
            return age > (hit.failed ? RETRY_MS : FRESH_MS);
        },

        forget(repo) {
            const all = held();
            delete all[keyFor(repo)];
            store(all);
        },

        forgetAll() {
            store({});
        },

        async refresh(repo) {
            if (!repo || !repo.token || !repo.repo || !GitApi || !GitApi.tree) return null;
            const key = keyFor(repo);
            if (pending.has(key)) return pending.get(key);
            const run = (async () => {
                try {
                    const read = await GitApi.tree(repo);
                    const kept = usable(read.paths, repo);
                    const all = held();
                    all[key] = {
                        at: Date.now(),
                        usedAt: Date.now(),
                        branch: read.branch || repo.branch || '',
                        paths: kept.paths,
                        dropped: kept.dropped,
                        total: kept.total,
                        partial: !!read.partial
                    };
                    keep(all);
                    return all[key];
                } catch (_) {
                    const all = held();
                    const before = all[key];
                    if (!before || !before.paths.length) {
                        all[key] = {
                            at: Date.now(),
                            usedAt: Date.now(),
                            branch: repo.branch || '',
                            paths: [],
                            dropped: 0,
                            total: 0,
                            partial: false,
                            failed: true
                        };
                        keep(all);
                    }
                    return null;
                } finally {
                    pending.delete(key);
                }
            })();
            pending.set(key, run);
            return run;
        },

        warm(repos) {
            for (const repo of repos || []) {
                if (this.stale(repo)) this.refresh(repo);
            }
        },

        touch(repos) {
            const all = held();
            const now = Date.now();
            let changed = false;
            for (const repo of repos || []) {
                const hit = all[keyFor(repo)];
                if (!hit) continue;
                if (now - (hit.usedAt || 0) < 60000) continue;
                hit.usedAt = now;
                changed = true;
            }
            if (changed) store(all);
        },

        async ready(repos, ms) {
            const list = (repos || []).filter(r => r && r.repo && r.token);
            this.touch(list);
            const missing = list.filter(r => {
                const hit = this.entry(r);
                return !hit || (!hit.paths.length && !hit.failed);
            });
            this.warm(list);
            if (!missing.length) return;
            const waits = missing.map(r => pending.get(keyFor(r))).filter(Boolean);
            if (!waits.length) return;
            await Promise.race([
                Promise.all(waits),
                new Promise(r => setTimeout(r, ms == null ? READY_MS : ms))
            ]);
        },

        block(repos, query) {
            const list = (repos || []).filter(r => r && r.repo);
            if (!list.length) return '';
            const terms = termsOf(query);
            const all = held();
            const sig = list.map(r => {
                const key = keyFor(r);
                const hit = all[key];
                return key + ':' + (hit ? hit.at + ':' + hit.paths.length : '0');
            }).join('|') + '|' + Array.from(terms).sort().join(',');
            if (sig === lastBlock.sig) return lastBlock.text;
            const parts = [];
            const budget = Math.floor(BLOCK_MAX / list.length);
            for (const repo of list) {
                const hit = all[keyFor(repo)];
                if (!hit || !Array.isArray(hit.paths) || !hit.paths.length) continue;
                parts.push(render(repo, hit, budget, terms));
            }
            const text = parts.length ? headed(parts) : '';
            lastBlock = { sig, text };
            return text;
        },

        usable,
        render,
        ignorable,
        group,
        BLOCK_MAX,
        FILES_PER_REPO,
        FRESH_MS
    };

    function headed(parts) {
        return '[repository files]\n'
                + 'What each repository in scope holds, on the branch named, read from the '
                + 'forge by this device just now. Each line is a directory, then the files '
                + 'in it — pictures, video and other assets included, since what a '
                + 'repository holds is part of the answer even when the file cannot be read '
                + 'as text. Go straight to the file you want rather than listing '
                + 'directories to find it. Only dependencies, build output and lockfiles '
                + 'are left out; a count stands in wherever names are not listed, and the '
                + 'listing can be a little behind the branch — so a file you expect and do '
                + 'not see here may still be there. Where a shorter or truncated file list '
                + 'appears elsewhere in your instructions, prefer this one.\n'
                + parts.join('\n');
    }

    window.NymbotRepoMap = RepoMap;
})();

File: flutter/lib/services/repo_map.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../models/workspace.dart';
import '../state/store.dart';
import 'git_forge.dart';

class RepoMapEntry {
  RepoMapEntry({
    required this.at,
    required this.usedAt,
    required this.branch,
    required this.paths,
    this.dropped = 0,
    this.total = 0,
    this.partial = false,
    this.failed = false,
  });

  final int at;
  int usedAt;
  final String branch;
  final List<String> paths;
  final int dropped;
  final int total;
  final bool partial;
  final bool failed;

  Map<String, dynamic> toJson() => {
        'at': at,
        'usedAt': usedAt,
        'branch': branch,
        'paths': paths,
        'dropped': dropped,
        'total': total,
        'partial': partial,
        'failed': failed,
      };

  static RepoMapEntry fromJson(Map<String, dynamic> j) => RepoMapEntry(
        at: (j['at'] as num?)?.toInt() ?? 0,
        usedAt: (j['usedAt'] as num?)?.toInt() ?? 0,
        branch: j['branch'] as String? ?? '',
        paths: (j['paths'] as List?)?.whereType<String>().toList() ?? const [],
        dropped: (j['dropped'] as num?)?.toInt() ?? 0,
        total: (j['total'] as num?)?.toInt() ?? 0,
        partial: j['partial'] == true,
        failed: j['failed'] == true,
      );
}

class RepoMap {
  RepoMap(this.store, {this.client});

  final Store store;
  final http.Client? client;

  static const cacheKey = 'repoMaps';
  static const freshFor = Duration(hours: 2);
  static const retryAfter = Duration(minutes: 5);
  static const keepMaps = 12;
  static const filesPerRepo = 500;
  static const dirNamesMax = 24;
  static const blockMax = 12000;
  static const readyWithin = Duration(milliseconds: 2500);

  static const _skipDirs = {
    'node_modules', '.git', 'dist', 'build', 'out', 'target', '.next',
    '.nuxt', '.svelte-kit', '.cache', '.parcel-cache', 'coverage',
    '__pycache__', '.venv', 'venv', 'Pods', '.dart_tool', '.gradle',
    '.idea', '.vscode', 'bower_components', '.terraform', 'DerivedData',
    '.pub-cache', '.mypy_cache', '.pytest_cache', '.tox', '.expo', '.angular',
  };

  static const _skipFiles = {
    'package-lock.json', 'yarn.lock', 'pnpm-lock.yaml', 'npm-shrinkwrap.json',
    'pubspec.lock', 'Cargo.lock', 'composer.lock', 'Gemfile.lock',
    'poetry.lock', 'go.sum', 'Podfile.lock', 'mix.lock', 'flake.lock',
    '.DS_Store',
  };

  static const _stopTerms = {
    'a', 'an', 'and', 'are', 'as', 'at', 'be', 'but', 'by', 'can', 'could',
    'did', 'do', 'does', 'file', 'files', 'for', 'from', 'had', 'has', 'have',
    'how', 'i', 'if', 'in', 'is', 'it', 'its', 'me', 'my', 'not', 'of', 'on',
    'or', 'our', 'please', 'so', 'than', 'that', 'the', 'their', 'them',
    'then', 'there', 'these', 'they', 'this', 'to', 'was', 'we', 'were',
    'what', 'when', 'where', 'which', 'who', 'why', 'will', 'with', 'would',
    'you', 'your',
  };

  final Map<String, Future<void>> _running = {};
  Map<String, RepoMapEntry>? _memo;
  String _blockSig = '';
  String _blockText = '';

  static String keyFor(GitRepo repo) => [
        repo.provider.isEmpty ? 'github' : repo.provider,
        repo.host,
        repo.repo,
        repo.branch,
        repo.paths,
      ].join('|');

  Map<String, RepoMapEntry> _held() {
    final memo = _memo;
    if (memo != null) return memo;
    final out = <String, RepoMapEntry>{};
    final raw = store.getString(cacheKey);
    if (raw != null && raw.isNotEmpty) {
      try {
        final decoded = jsonDecode(raw);
        if (decoded is Map) {
          decoded.forEach((k, v) {
            if (k is String && v is Map<String, dynamic>) {
              out[k] = RepoMapEntry.fromJson(v);
            }
          });
        }
      } catch (_) {
        out.clear();
      }
    }
    _memo = out;
    return out;
  }

  Future<void> _store(Map<String, RepoMapEntry> all) {
    _memo = all;
    return store.setString(cacheKey,
        jsonEncode(all.map((k, v) => MapEntry(k, v.toJson()))));
  }

  Future<void> _keep(Map<String, RepoMapEntry> all) {
    if (all.length <= keepMaps) return _store(all);
    final keys = all.keys.toList()
      ..sort((a, b) => (all[b]!.usedAt == 0 ? all[b]!.at : all[b]!.usedAt)
          .compareTo(all[a]!.usedAt == 0 ? all[a]!.at : all[a]!.usedAt));
    final out = <String, RepoMapEntry>{};
    for (final k in keys.take(keepMaps)) {
      out[k] = all[k]!;
    }
    return _store(out);
  }

  static List<String> _scopes(GitRepo? repo) => (repo?.paths ?? '')
      .split(RegExp(r'[,\s]+'))
      .where((p) => p.isNotEmpty)
      .map((p) => p.replaceFirst(RegExp(r'^/+'), ''))
      .toList();

  static bool _inScope(String path, List<String> only) {
    if (only.isEmpty) return true;
    for (final p in only) {
      if (path == p) return true;
      if (path.startsWith(p.endsWith('/') ? p : '$p/')) return true;
    }
    return false;
  }

  static bool ignorable(String path) {
    final parts = path.split('/');
    if (_skipFiles.contains(parts.last)) return true;
    for (var i = 0; i < parts.length - 1; i++) {
      if (_skipDirs.contains(parts[i])) return true;
    }
    return false;
  }

  static int _depthOf(String path) {
    var n = 0;
    for (var i = 0; i < path.length; i++) {
      if (path[i] == '/') n++;
    }
    return n;
  }

  static int _byName(String a, String b) {
    final plain = a.toLowerCase().compareTo(b.toLowerCase());
    return plain != 0 ? plain : a.compareTo(b);
  }

  static ({List<String> paths, int dropped, int total}) usable(
      List<String> files, GitRepo? repo) {
    final only = _scopes(repo);
    final rows = <String>[];
    for (final path in files) {
      if (path.isEmpty || path.endsWith('/')) continue;
      if (ignorable(path)) continue;
      if (!_inScope(path, only)) continue;
      rows.add(path);
    }
    rows.sort((a, b) {
      final depth = _depthOf(a).compareTo(_depthOf(b));
      return depth != 0 ? depth : _byName(a, b);
    });
    final kept = rows.take(filesPerRepo).toList()..sort(_byName);
    return (paths: kept, dropped: rows.length - kept.length, total: rows.length);
  }

  static Set<String> _termsOf(String query) => query
      .toLowerCase()
      .split(RegExp(r'[^a-z0-9]+'))
      .where((w) => w.length > 2 && !_stopTerms.contains(w))
      .toSet();

  static bool _wanted(String name, Set<String> terms) {
    if (terms.isEmpty) return false;
    final plain = name.toLowerCase();
    for (final term in terms) {
      if (plain.contains(term)) return true;
    }
    return false;
  }

  static Map<String, List<String>> group(List<String> paths) {
    final dirs = <String, List<String>>{};
    for (final path in paths) {
      final at = path.lastIndexOf('/');
      final dir = at == -1 ? '' : path.substring(0, at + 1);
      final name = at == -1 ? path : path.substring(at + 1);
      dirs.putIfAbsent(dir, () => <String>[]).add(name);
    }
    return dirs;
  }

  static String render(
      String repo, RepoMapEntry entry, int budget, Set<String> terms,
      {String paths = ''}) {
    final branch = entry.branch.isEmpty ? '' : '@${entry.branch}';
    final scope = paths.isEmpty ? '' : ', paths $paths';
    final plural = entry.total == 1 ? ' file' : ' files';
    final lines = <String>['--- $repo$branch (${entry.total}$plural$scope) ---'];
    var spent = lines.first.length;
    final dirs = group(entry.paths);
    final names = dirs.keys.toList()..sort();
    var skipped = 0;

    for (final dir in names) {
      final all = dirs[dir]!..sort(_byName);
      if (spent >= budget) {
        skipped += all.length;
        continue;
      }
      final ordered = [
        ...all.where((n) => _wanted(n, terms)),
        ...all.where((n) => !_wanted(n, terms)),
      ];
      final shown = ordered.take(dirNamesMax).toList();
      final over = ordered.length - shown.length;
      final line = '${dir.isEmpty ? '(root)' : dir}: ${shown.join(', ')}'
          '${over > 0 ? ', and $over more here' : ''}';
      spent += line.length + 1;
      lines.add(line);
    }

    final unlisted = entry.dropped + skipped;
    if (unlisted > 0) {
      lines.add('$unlisted further files are not named above '
          '— list a directory to see them.');
    }
    if (entry.partial) {
      lines.add('The forge cut this listing short, so it is not every file.');
    }
    return lines.join('\n');
  }

  RepoMapEntry? entry(GitRepo repo) => _held()[keyFor(repo)];

  bool stale(GitRepo repo) {
    final hit = entry(repo);
    if (hit == null) return true;
    final age = DateTime.now().millisecondsSinceEpoch - hit.at;
    return age > (hit.failed ? retryAfter : freshFor).inMilliseconds;
  }

  Future<void> forget(GitRepo repo) {
    final all = _held();
    all.remove(keyFor(repo));
    return _store(all);
  }

  Future<void> forgetAll() => _store(<String, RepoMapEntry>{});

  Future<void> refresh(GitRepo repo) {
    if (repo.token.isEmpty || repo.repo.isEmpty) return Future.value();
    final key = keyFor(repo);
    final already = _running[key];
    if (already != null) return already;
    final run = () async {
      final now = DateTime.now().millisecondsSinceEpoch;
      try {
        final read = await GitForge.tree(
          provider: repo.provider.isEmpty ? 'github' : repo.provider,
          token: repo.token,
          repo: repo.repo,
          host: repo.host,
          branch: repo.branch,
          client: client,
        );
        final kept = usable(read.paths, repo);
        final all = _held();
        all[key] = RepoMapEntry(
          at: now,
          usedAt: now,
          branch: read.branch.isEmpty ? repo.branch : read.branch,
          paths: kept.paths,
          dropped: kept.dropped,
          total: kept.total,
          partial: read.partial,
        );
        await _keep(all);
      } catch (_) {
        final all = _held();
        final before = all[key];
        if (before == null || before.paths.isEmpty) {
          all[key] = RepoMapEntry(
            at: now,
            usedAt: now,
            branch: repo.branch,
            paths: const [],
            failed: true,
          );
          await _keep(all);
        }
      } finally {
        _done(key);
      }
    }();
    _running[key] = run;
    return run;
  }

  void _done(String key) {
    _running.remove(key);
  }

  void warm(List<GitRepo> repos) {
    for (final repo in repos) {
      if (stale(repo)) refresh(repo);
    }
  }

  Future<void> touch(List<GitRepo> repos) {
    final all = _held();
    final now = DateTime.now().millisecondsSinceEpoch;
    var changed = false;
    for (final repo in repos) {
      final hit = all[keyFor(repo)];
      if (hit == null || now - hit.usedAt < 60000) continue;
      hit.usedAt = now;
      changed = true;
    }
    return changed ? _store(all) : Future.value();
  }

  Future<void> ready(List<GitRepo> repos, {Duration? within}) async {
    final list = repos.where((r) => r.repo.isNotEmpty && r.token.isNotEmpty).toList();
    await touch(list);
    final missing = list.where((r) {
      final hit = entry(r);
      return hit == null || (hit.paths.isEmpty && !hit.failed);
    }).toList();
    warm(list);
    if (missing.isEmpty) return;
    final waits = missing
        .map((r) => _running[keyFor(r)])
        .whereType<Future<void>>()
        .toList();
    if (waits.isEmpty) return;
    await Future.any([
      Future.wait(waits),
      Future<void>.delayed(within ?? readyWithin),
    ]);
  }

  bool knows(List<GitRepo> repos) =>
      repos.every((r) => entry(r) != null);

  String block(List<GitRepo> repos, String query) {
    final list = repos.where((r) => r.repo.isNotEmpty).toList();
    if (list.isEmpty) return '';
    final terms = _termsOf(query);
    final all = _held();
    final sig = '${list.map((r) {
      final key = keyFor(r);
      final hit = all[key];
      return '$key:${hit == null ? '0' : '${hit.at}:${hit.paths.length}'}';
    }).join('|')}|${(terms.toList()..sort()).join(',')}';
    if (sig == _blockSig) return _blockText;
    final parts = <String>[];
    final budget = blockMax ~/ list.length;
    for (final repo in list) {
      final hit = all[keyFor(repo)];
      if (hit == null || hit.paths.isEmpty) continue;
      parts.add(render(repo.repo, hit, budget, terms, paths: repo.paths));
    }
    _blockSig = sig;
    _blockText = parts.isEmpty ? '' : _headed(parts);
    return _blockText;
  }

  static String _headed(List<String> parts) {
    return '[repository files]\n'
        'What each repository in scope holds, on the branch named, read from the '
        'forge by this device just now. Each line is a directory, then the files '
        'in it — pictures, video and other assets included, since what a '
        'repository holds is part of the answer even when the file cannot be read '
        'as text. Go straight to the file you want rather than listing '
        'directories to find it. Only dependencies, build output and lockfiles '
        'are left out; a count stands in wherever names are not listed, and the '
        'listing can be a little behind the branch — so a file you expect and do '
        'not see here may still be there. Where a shorter or truncated file list '
        'appears elsewhere in your instructions, prefer this one.\n'
        '${parts.join('\n')}';
  }
}
